protocols/rendezvous/src/server.rs

```diff
@@ -142,187 +151,179 @@ impl NetworkBehaviour for Behaviour {
+        Message::Unregister(namespace) => {
+            registrations.remove(namespace.clone(), peer_id);
+
+            let event = Event::PeerUnregistered {
+                peer: peer_id,
+                namespace,
+            };
+
+            (event, None)
+        }
+        Message::Discover {
             namespace,
             cookie,
             limit,
         } => match registrations.get(namespace, cookie, limit) {
             Ok((registrations, cookie)) => {
                 let discovered = registrations.cloned().collect::<Vec<_>>();
 
-                vec![
-                    ToSwarm::NotifyHandler {
-                        peer_id,
-                        handler: NotifyHandler::One(connection),
-                        event: handler::InboundInEvent::NotifyInboundSubstream {
-                            id,
-                            message: inbound::InEvent::DiscoverResponse {
-                                discovered: discovered.clone(),
-                                cookie,
-                            },
-                        },
-                    },
-                    ToSwarm::GenerateEvent(Event::DiscoverServed {
-                        enquirer: peer_id,
-                        registrations: discovered,
-                    }),
-                ]
+                let response = Message::DiscoverResponse(Ok((discovered.clone(), cookie)));
+
+                let event = Event::DiscoverServed {
+                    enquirer: peer_id,
+                    registrations: discovered,
+                };
+
+                (event, Some(response))
             }
             Err(_) => {
                 let error = ErrorCode::InvalidCookie;
 
-                vec![
-                    ToSwarm::NotifyHandler {
-                        peer_id,
-                        handler: NotifyHandler::One(connection),
-                        event: handler::InboundInEvent::NotifyInboundSubstream {
-                            id,
-                            message: inbound::InEvent::DeclineDiscoverRequest(error),
-                        },
-                    },
-                    ToSwarm::GenerateEvent(Event::DiscoverNotServed {
-                        enquirer: peer_id,
-                        error,
-                    }),
-                ]
+                let response = Message::DiscoverResponse(Err(error));
+
+                let event = Event::DiscoverNotServed {
+                    enquirer: peer_id,
+                    error,
+                };
+
+                (event, Some(response))
             }
         },
-        inbound::OutEvent::UnregisterRequested(namespace) => {
-            registrations.remove(namespace.clone(), peer_id);
-
-            vec![ToSwarm::GenerateEvent(Event::PeerUnregistered {
-                peer: peer_id,
-                namespace,
-            })]
+        Message::RegisterResponse(_) => {
+            unreachable!()
```

Comment:
These could be triggered by a malicious client if they violate the protocol and send us "responses" as requests.
We will have to return `None` here and `continue` in the outer loop.